File: impl/csharp/Cbor.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace Naalp
{
    /// <summary>
    /// Deterministic CBOR (RFC 8949 §4.2.1) for N-AALP — the C1 codec.
    ///
    /// <para>An independent C# implementation of the same deterministic profile the Go, Rust, Python
    /// and Java reference implementations produce: shortest-form integer heads, no indefinite lengths,
    /// canonical (bytewise-ascending, by encoded key) map ordering, and no duplicate keys. The strict
    /// decoder rejects every non-canonical encoding — out-of-order or duplicate map keys, non-shortest
    /// integers, indefinite lengths, trailing bytes — with a NaalpException of kind NonCanonical. The
    /// content id is multihash(0x20, 0x30, SHA-384(body)) over the deterministic body bytes (§2.3).</para>
    /// </summary>
    public static class Cbor
    {
        // Strict UTF-8: matches Python's str.decode("utf-8") / encode("utf-8") — reject invalid bytes,
        // never emit a BOM.
        private static readonly UTF8Encoding Utf8 = new UTF8Encoding(false, true);

        // --- value model (mirrors the Go/Rust/Python/Java cbor.Value variants) ---

        /// <summary>Base of the CBOR value hierarchy.</summary>
        public abstract class Value { }

        /// <summary>CBOR unsigned integer (major 0).</summary>
        public sealed class U : Value
        {
            public readonly long V;
            public U(long v) { V = v; }
        }

        /// <summary>CBOR negative integer (major 1); V is the negative value itself.</summary>
        public sealed class N : Value
        {
            public readonly long V;
            public N(long v) { V = v; }
        }

        /// <summary>CBOR byte string (major 2).</summary>
        public sealed class B : Value
        {
            public readonly byte[] V;
            public B(byte[] v) { V = v; }
        }

        /// <summary>CBOR text string (major 3).</summary>
        public sealed class T : Value
        {
            public readonly string V;
            public T(string v) { V = v; }
        }

        /// <summary>CBOR array (major 4).</summary>
        public sealed class A : Value
        {
            public readonly List<Value> Items;
            public A(List<Value> items) { Items = items; }
        }

        /// <summary>A single map entry.</summary>
        public sealed class Pair
        {
            public readonly Value K;
            public readonly Value Val;
            public Pair(Value k, Value val) { K = k; Val = val; }
        }

        /// <summary>CBOR map (major 5); Pairs is a list of (key, value).</summary>
        public sealed class M : Value
        {
            public readonly List<Pair> Pairs;
            public M(List<Pair> pairs) { Pairs = pairs; }
        }

        /// <summary>CBOR tag (major 6).</summary>
        public sealed class Tag : Value
        {
            public readonly long N;
            public readonly Value Content;
            public Tag(long n, Value content) { N = n; Content = content; }
        }

        // --- encoder ---

        private static byte[] Head(int major, long n)
        {
            // All call sites pass a non-negative n; use unsigned arithmetic for the head so shifts and
            // the shortest-form thresholds are exact even for large 64-bit values.
            ulong u = (ulong)n;
            int m = major << 5;
            if (u < 24)
            {
                return new byte[] { (byte)(m | (int)u) };
            }
            if (u < 256)
            {
                return new byte[] { (byte)(m | 24), (byte)u };
            }
            if (u < 65536UL)
            {
                return new byte[] { (byte)(m | 25), (byte)(u >> 8), (byte)u };
            }
            if (u < 4294967296UL)
            {
                return new byte[]
                {
                    (byte)(m | 26),
                    (byte)(u >> 24), (byte)(u >> 16), (byte)(u >> 8), (byte)u,
                };
            }
            return new byte[]
            {
                (byte)(m | 27),
                (byte)(u >> 56), (byte)(u >> 48), (byte)(u >> 40), (byte)(u >> 32),
                (byte)(u >> 24), (byte)(u >> 16), (byte)(u >> 8), (byte)u,
            };
        }

        /// <summary>Deterministic-CBOR encode a value; map keys are emitted in canonical order.</summary>
        public static byte[] Encode(Value v)
        {
            var outp = new MemoryStream();
            EncodeInto(v, outp);
            return outp.ToArray();
        }

        private static void Write(MemoryStream outp, byte[] b)
        {
            outp.Write(b, 0, b.Length);
        }

        private static void EncodeInto(Value v, MemoryStream outp)
        {
            switch (v)
            {
                case U u:
                    if (u.V < 0)
                    {
                        throw new NaalpException("NonCanonical", "uint is negative");
                    }
                    Write(outp, Head(0, u.V));
                    break;
                case N n:
                    Write(outp, Head(1, -1 - n.V));
                    break;
                case B b:
                    Write(outp, Head(2, b.V.Length));
                    Write(outp, b.V);
                    break;
                case T t:
                    byte[] s = Utf8.GetBytes(t.V);
                    Write(outp, Head(3, s.Length));
                    Write(outp, s);
                    break;
                case A a:
                    Write(outp, Head(4, a.Items.Count));
                    foreach (Value item in a.Items)
                    {
                        EncodeInto(item, outp);
                    }
                    break;
                case M m:
                {
                    int count = m.Pairs.Count;
                    var keys = new byte[count][];
                    var vals = new byte[count][];
                    for (int i = 0; i < count; i++)
                    {
                        keys[i] = Encode(m.Pairs[i].K);
                        vals[i] = Encode(m.Pairs[i].Val);
                    }
                    var order = new int[count];
                    for (int i = 0; i < count; i++)
                    {
                        order[i] = i;
                    }
                    Array.Sort(order, (x, y) => CompareBytes(keys[x], keys[y]));
                    Write(outp, Head(5, count));
                    byte[]? prev = null;
                    foreach (int idx in order)
                    {
                        byte[] k = keys[idx];
                        if (prev != null && CompareBytes(prev, k) == 0)
                        {
                            throw new NaalpException("NonCanonical", "duplicate map key");
                        }
                        prev = k;
                        Write(outp, k);
                        Write(outp, vals[idx]);
                    }
                    break;
                }
                case Tag tg:
                    Write(outp, Head(6, tg.N));
                    EncodeInto(tg.Content, outp);
                    break;
                default:
                    throw new NaalpException("NonCanonical", "not a cbor value");
            }
        }

        /// <summary>Unsigned bytewise lexicographic comparison.</summary>
        internal static int CompareBytes(byte[] a, byte[] b)
        {
            int n = Math.Min(a.Length, b.Length);
            for (int i = 0; i < n; i++)
            {
                int x = a[i] & 0xFF;
                int y = b[i] & 0xFF;
                if (x != y)
                {
                    return x - y;
                }
            }
            return a.Length - b.Length;
        }

        // --- decoder (strict canonical) ---

        private sealed class Cursor
        {
            public readonly byte[] Data;
            public int Pos;
            public Cursor(byte[] data) { Data = data; }
            public int Remaining => Data.Length - Pos;
        }

        private static Value Dec(Cursor c)
        {
            if (c.Remaining < 1)
            {
                throw new NaalpException("NonCanonical", "truncated");
            }
            int ib = c.Data[c.Pos++] & 0xFF;
            int major = ib >> 5;
            int ai = ib & 0x1F;
            if (ai == 31)
            {
                throw new NaalpException("NonCanonical", "indefinite length");
            }
            ulong arg;
            if (ai < 24)
            {
                arg = (ulong)ai;
            }
            else if (ai == 24)
            {
                if (c.Remaining < 1)
                {
                    throw new NaalpException("NonCanonical", "truncated head");
                }
                arg = c.Data[c.Pos++] & 0xFFUL;
                if (arg < 24)
                {
                    throw new NaalpException("NonCanonical", "non-shortest integer");
                }
            }
            else if (ai == 25)
            {
                if (c.Remaining < 2)
                {
                    throw new NaalpException("NonCanonical", "truncated head");
                }
                arg = ReadBE(c, 2);
                if (arg < 256)
                {
                    throw new NaalpException("NonCanonical", "non-shortest integer");
                }
            }
            else if (ai == 26)
            {
                if (c.Remaining < 4)
                {
                    throw new NaalpException("NonCanonical", "truncated head");
                }
                arg = ReadBE(c, 4);
                if (arg < 65536UL)
                {
                    throw new NaalpException("NonCanonical", "non-shortest integer");
                }
            }
            else if (ai == 27)
            {
                if (c.Remaining < 8)
                {
                    throw new NaalpException("NonCanonical", "truncated head");
                }
                arg = ReadBE(c, 8);
                if (arg < 4294967296UL)
                {
                    throw new NaalpException("NonCanonical", "non-shortest integer");
                }
            }
            else
            {
                throw new NaalpException("NonCanonical", "reserved additional-info");
            }

            switch (major)
            {
                case 0:
                    return new U((long)arg);
                case 1:
                    return new N(-1L - (long)arg);
                case 2:
                {
                    int len = LenOf(arg);
                    if (c.Remaining < len)
                    {
                        throw new NaalpException("NonCanonical", "truncated byte string");
                    }
                    byte[] b = new byte[len];
                    Array.Copy(c.Data, c.Pos, b, 0, len);
                    c.Pos += len;
                    return new B(b);
                }
                case 3:
                {
                    int len = LenOf(arg);
                    if (c.Remaining < len)
                    {
                        throw new NaalpException("NonCanonical", "truncated text string");
                    }
                    string s;
                    try
                    {
                        s = Utf8.GetString(c.Data, c.Pos, len);
                    }
                    catch (Exception)
                    {
                        throw new NaalpException("NonCanonical", "text string is not valid UTF-8");
                    }
                    c.Pos += len;
                    return new T(s);
                }
                case 4:
                {
                    int len = LenOf(arg);
                    var items = new List<Value>(len);
                    for (int i = 0; i < len; i++)
                    {
                        items.Add(Dec(c));
                    }
                    return new A(items);
                }
                case 5:
                {
                    int len = LenOf(arg);
                    var pairs = new List<Pair>(len);
                    byte[]? prev = null;
                    for (int i = 0; i < len; i++)
                    {
                        int before = c.Pos;
                        Value k = Dec(c);
                        byte[] kbytes = new byte[c.Pos - before];
                        Array.Copy(c.Data, before, kbytes, 0, kbytes.Length);
                        Value val = Dec(c);
                        if (prev != null && CompareBytes(kbytes, prev) <= 0)
                        {
                            throw new NaalpException("NonCanonical", "map keys out of order or duplicate");
                        }
                        prev = kbytes;
                        pairs.Add(new Pair(k, val));
                    }
                    return new M(pairs);
                }
                case 6:
                {
                    Value content = Dec(c);
                    return new Tag((long)arg, content);
                }
                default:
                    throw new NaalpException("NonCanonical", "unsupported major type " + major);
            }
        }

        private static ulong ReadBE(Cursor c, int n)
        {
            ulong v = 0;
            for (int i = 0; i < n; i++)
            {
                v = (v << 8) | (c.Data[c.Pos++] & 0xFFUL);
            }
            return v;
        }

        private static int LenOf(ulong arg)
        {
            if (arg > int.MaxValue)
            {
                throw new NaalpException("NonCanonical", "length out of range");
            }
            return (int)arg;
        }

        /// <summary>Strict canonical decode: rejects any non-canonical encoding with a NonCanonical error.</summary>
        public static Value Decode(byte[] data)
        {
            var c = new Cursor(data);
            Value v = Dec(c);
            if (c.Remaining != 0)
            {
                throw new NaalpException("NonCanonical", "trailing bytes after top-level item");
            }
            return v;
        }

        // --- content id ---

        /// <summary>Content id = multihash(0x20 [sha2-384], 0x30 [48], SHA-384(body-bytes)) (§2.3).</summary>
        public static byte[] ContentId(byte[] body)
        {
            byte[] digest;
            using (var sha = SHA384.Create())
            {
                digest = sha.ComputeHash(body);
            }
            byte[] outp = new byte[2 + digest.Length];
            outp[0] = 0x20;
            outp[1] = 0x30;
            Array.Copy(digest, 0, outp, 2, digest.Length);
            return outp;
        }

        /// <summary>Content id over a value (encode then hash).</summary>
        public static byte[] ContentId(Value v)
        {
            return ContentId(Encode(v));
        }
    }
}

File: impl/php/src/Cbor.php
<?php
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
/**
 * Deterministic CBOR (RFC 8949 §4.2.1) for N-AALP — the C1 codec.
 *
 * An independent PHP port of the same deterministic profile the Python, Go, and Rust
 * reference implementations produce: shortest-form integer heads, no indefinite lengths,
 * canonical (bytewise-ascending, by encoded key) map ordering, no duplicate keys. The
 * strict decoder rejects every non-canonical form. content_id is
 * multihash(0x20 sha2-384, 0x30 len-48, SHA-384(body)) over the deterministic body bytes (§2.3).
 *
 * The value model (U/N/B/T/A/M/Tag) mirrors the Python cbor.Value variants exactly so the
 * emitted bytes are byte-identical.
 */

declare(strict_types=1);

namespace Naalp;

/** Raised on any non-canonical CBOR (encode or strict decode). */
class NonCanonical extends \RuntimeException
{
    public string $kind = "NonCanonical";
}

// --- value model (mirrors the Go/Rust/Python cbor.Value variants) ---

/** CBOR unsigned integer (major 0). */
final class U
{
    public int $v;
    public function __construct(int $v) { $this->v = $v; }
}

/** CBOR negative integer (major 1); v is the negative value itself. */
final class N
{
    public int $v;
    public function __construct(int $v) { $this->v = $v; }
}

/** CBOR byte string (major 2); v is a raw binary string. */
final class B
{
    public string $v;
    public function __construct(string $v) { $this->v = $v; }
}

/** CBOR text string (major 3); v is a UTF-8 string. */
final class T
{
    public string $v;
    public function __construct(string $v) { $this->v = $v; }
}

/** CBOR array (major 4). */
final class A
{
    /** @var array<int,mixed> */
    public array $items;
    /** @param array<int,mixed> $items */
    public function __construct(array $items) { $this->items = array_values($items); }
}

/** CBOR map (major 5); pairs is a list of [key_value, value]. */
final class M
{
    /** @var array<int,array{0:mixed,1:mixed}> */
    public array $pairs;
    /** @param array<int,array{0:mixed,1:mixed}> $pairs */
    public function __construct(array $pairs) { $this->pairs = array_values($pairs); }
}

/** CBOR tag (major 6). */
final class Tag
{
    public int $n;
    public mixed $content;
    public function __construct(int $n, mixed $content) { $this->n = $n; $this->content = $content; }
}

final class Cbor
{
    /** Encode a definite-length head: (major<<5)|arg, shortest form (RFC 8949 §4.2.1). */
    private static function head(int $major, int $n): string
    {
        $m = $major << 5;
        if ($n < 24) {
            return \chr($m | $n);
        }
        if ($n < 256) {
            return \chr($m | 24) . \chr($n);
        }
        if ($n < 65536) {
            return \chr($m | 25) . \pack('n', $n);
        }
        if ($n < 4294967296) {
            return \chr($m | 26) . \pack('N', $n);
        }
        return \chr($m | 27) . \pack('J', $n);
    }

    /** Deterministic-CBOR encode a value; map keys are emitted in canonical order. */
    public static function encode(mixed $v): string
    {
        if ($v instanceof U) {
            if ($v->v < 0) {
                throw new NonCanonical("uint is negative");
            }
            return self::head(0, $v->v);
        }
        if ($v instanceof N) {
            return self::head(1, -1 - $v->v);
        }
        if ($v instanceof B) {
            return self::head(2, \strlen($v->v)) . $v->v;
        }
        if ($v instanceof T) {
            return self::head(3, \strlen($v->v)) . $v->v;
        }
        if ($v instanceof A) {
            $out = self::head(4, \count($v->items));
            foreach ($v->items as $i) {
                $out .= self::encode($i);
            }
            return $out;
        }
        if ($v instanceof M) {
            $enc = [];
            foreach ($v->pairs as $kv) {
                $enc[] = [self::encode($kv[0]), self::encode($kv[1])];
            }
            // canonical: sort by encoded-key bytes, bytewise ascending (strcmp = memcmp, unsigned)
            \usort($enc, static fn(array $a, array $b): int => \strcmp($a[0], $b[0]));
            for ($i = 1; $i < \count($enc); $i++) {
                if ($enc[$i][0] === $enc[$i - 1][0]) {
                    throw new NonCanonical("duplicate map key");
                }
            }
            $out = self::head(5, \count($enc));
            foreach ($enc as $kv) {
                $out .= $kv[0] . $kv[1];
            }
            return $out;
        }
        if ($v instanceof Tag) {
            return self::head(6, $v->n) . self::encode($v->content);
        }
        throw new \TypeError("not a cbor value");
    }

    /**
     * Decode one item at $off. Returns [value, newOffset]. Strict: rejects any non-canonical
     * encoding (non-shortest int, indefinite length, out-of-order/duplicate map keys, trailing bytes).
     *
     * @return array{0:mixed,1:int}
     */
    private static function dec(string $data, int $off): array
    {
        $len = \strlen($data);
        if ($off >= $len) {
            throw new NonCanonical("truncated");
        }
        $ib = \ord($data[$off]);
        $major = $ib >> 5;
        $ai = $ib & 0x1F;
        $p = $off + 1;

        if ($ai === 31) {
            throw new NonCanonical("indefinite length");
        }
        if ($ai < 24) {
            $arg = $ai;
        } elseif ($ai === 24) {
            if ($p + 1 > $len) {
                throw new NonCanonical("truncated head");
            }
            $arg = \ord($data[$p]);
            $p += 1;
            if ($arg < 24) {
                throw new NonCanonical("non-shortest integer");
            }
        } elseif ($ai === 25) {
            if ($p + 2 > $len) {
                throw new NonCanonical("truncated head");
            }
            $arg = (\ord($data[$p]) << 8) | \ord($data[$p + 1]);
            $p += 2;
            if ($arg < 256) {
                throw new NonCanonical("non-shortest integer");
            }
        } elseif ($ai === 26) {
            if ($p + 4 > $len) {
                throw new NonCanonical("truncated head");
            }
            $arg = 0;
            for ($k = 0; $k < 4; $k++) {
                $arg = ($arg << 8) | \ord($data[$p + $k]);
            }
            $p += 4;
            if ($arg < 65536) {
                throw new NonCanonical("non-shortest integer");
            }
        } elseif ($ai === 27) {
            if ($p + 8 > $len) {
                throw new NonCanonical("truncated head");
            }
            // 64-bit big-endian; corpus values fit in a signed 64-bit int
            $u = \unpack('J', \substr($data, $p, 8));
            $arg = $u[1];
            $p += 8;
            if ($arg >= 0 && $arg < 4294967296) {
                throw new NonCanonical("non-shortest integer");
            }
        } else {
            throw new NonCanonical("reserved additional-info");
        }

        switch ($major) {
            case 0:
                return [new U($arg), $p];
            case 1:
                return [new N(-1 - $arg), $p];
            case 2:
                if ($p + $arg > $len) {
                    throw new NonCanonical("truncated byte string");
                }
                return [new B(\substr($data, $p, $arg)), $p + $arg];
            case 3:
                if ($p + $arg > $len) {
                    throw new NonCanonical("truncated text string");
                }
                return [new T(\substr($data, $p, $arg)), $p + $arg];
            case 4:
                $items = [];
                for ($i = 0; $i < $arg; $i++) {
                    [$it, $p] = self::dec($data, $p);
                    $items[] = $it;
                }
                return [new A($items), $p];
            case 5:
                $pairs = [];
                $prev = null;
                for ($i = 0; $i < $arg; $i++) {
                    $before = $p;
                    [$k, $p] = self::dec($data, $p);
                    $kbytes = \substr($data, $before, $p - $before);
                    [$val, $p] = self::dec($data, $p);
                    if ($prev !== null && \strcmp($kbytes, $prev) <= 0) {
                        throw new NonCanonical("map keys out of order or duplicate");
                    }
                    $prev = $kbytes;
                    $pairs[] = [$k, $val];
                }
                return [new M($pairs), $p];
            case 6:
                [$content, $p] = self::dec($data, $p);
                return [new Tag($arg, $content), $p];
            default:
                throw new NonCanonical("unsupported major type");
        }
    }

    /** Strict canonical decode: rejects any non-canonical encoding with NonCanonical. */
    public static function decode(string $data): mixed
    {
        [$v, $off] = self::dec($data, 0);
        if ($off !== \strlen($data)) {
            throw new NonCanonical("trailing bytes after top-level item");
        }
        return $v;
    }

    /**
     * content id = multihash(0x20 [sha2-384], 0x30 [48], SHA-384(body-bytes)) (§2.3).
     * $body may be a cbor value (re-encoded) or a raw binary string.
     */
    public static function contentId(mixed $body): string
    {
        if ($body instanceof U || $body instanceof N || $body instanceof B
            || $body instanceof T || $body instanceof A || $body instanceof M || $body instanceof Tag) {
            $body = self::encode($body);
        }
        return "\x20\x30" . \hash('sha384', (string) $body, true);
    }
}
